### PR TITLE
InstallBasicPackagesFiles improvements

### PR DESCRIPTION
**Pull request by drdanz**
* Add ``TARGETS`` argument to allow passing the targets without using a property
* Add ``DEPENDENCIES`` argument.
  Dependencies added to this list will be searched using ``find_dependency``.
  The dependencies will replace the ``@PACKAGE_DEPENDENCIES@`` string in the template file.
* Fixed namespace in exported targets.
* Keep internal files in current directory.
* Add ``CONFIG_TEMPLATE`` argument.
  Also allow the config file in current source directory and improve ``UPPERCASE_FILENAMES`` and ``LOWERCASE_FILENAME`` handling.
* Allow to use ``@PACKAGE_VERSION@`` in template files.
  This argument is replaced with the value of the ``VERSION`` argument.
* Add ``NO_COMPATIBILITY_VARS`` arguments.
  If set, the generation of the compatibility variables ``<VARS_PREFIX>_LIBRARIES`` and ``<VARS_PREFIX>_INCLUDE_DIRS`` is disabled.
  Also improve the generation of the ``<VARS_PREFIX>_INCLUDE_DIRS`` variable.

### CODE REVIEW

**Comment by drdanz** [comment]
Most of the changes are improvements, but since some time ago we had some compatibility discussion when we changed from `set_and_check()` to `set()`, and I don't remember exactly what agreement we reached, I'd like to be sure that we agree on this...

The original code was this:

```cmake
set_and_check(${_IBPF_VARS_PREFIX}_INCLUDEDIR \"@PACKAGE_${_IBPF_VARS_PREFIX}_INCLUDEDIR@\")
```

later, since the `set_and_check()` was having issues when `INCLUDEDIR` was not set, it was changed to this:

```cmake
set(${_IBPF_VARS_PREFIX}_INCLUDEDIR \"@PACKAGE_${_IBPF_VARS_PREFIX}_INCLUDEDIR@\")
```

In the new version I made this way more complicated, in order to support also `INCLUDE_DIR` and if none is set, to try to extract the variable from the target, and it became a piece of code that is generated by this check:

```cmake
if(DEFINED ${_IBPF_VARS_PREFIX}_BUILD_INCLUDEDIR OR
   DEFINED BUILD_${_IBPF_VARS_PREFIX}_INCLUDEDIR OR
   DEFINED ${_IBPF_VARS_PREFIX}_INSTALL_INCLUDEDIR OR
   DEFINED INSTALL_${_IBPF_VARS_PREFIX}_INCLUDEDIR)
  set(_get_include_dir "set(${_IBPF_VARS_PREFIX}_INCLUDEDIR \"\@PACKAGE_${_IBPF_VARS_PREFIX}_INCLUDEDIR\@\")")
elseif(DEFINED ${_IBPF_VARS_PREFIX}_BUILD_INCLUDE_DIR OR
       DEFINED BUILD_${_IBPF_VARS_PREFIX}_INCLUDE_DIR OR
       DEFINED ${_IBPF_VARS_PREFIX}_INSTALL_INCLUDE_DIR OR
       DEFINED INSTALL_${_IBPF_VARS_PREFIX}_INCLUDE_DIR)
  string(APPEND _get_include_dir "set(${_IBPF_VARS_PREFIX}_INCLUDE_DIR \"\@PACKAGE_${_IBPF_VARS_PREFIX}_INCLUDE_DIR\@\")")
else()
  foreach(_target ${${_IBPF_VARS_PREFIX}_TARGETS})
    string(APPEND _get_include_dir "get_property(${_IBPF_VARS_PREFIX}_INCLUDE_DIR TARGET ${_target} PROPERTY INTERFACE_INCLUDE_DIRECTORIES)")
  endforeach()
endif()
```

Does this look ok to you? CC @traversaro @claudiofantacci

**Comment by claudiofantacci** [comment]
> Most of the changes are improvements

Thanks for this PR, it seems to me that this function has been improved A LOT 🎉 !
I just went through it, but I'm not 100% sure of the changes because I honestly don't have so much "low level" expertise to assess whether something is good or not.
On the other hand, what I can do is try and test it on my libraries and give you feedback.
Is that ok for yo 😄 ?

> we changed from set_and_check() to set(), and I don't remember exactly what agreement we reached, I'd like to be sure that we agree on this

I'm on your side, I don't remember 100% this, but I recall the following: when installing a package with `InstallBasicPackagesFiles` with a new version of CMake (I cannot recall from which one) the `<package>_INCLUDE_DIR` was not exported and the `set_and_check()` was returning an error.
Your snipped code seems to account this and honestly I don't think is _that_ complicated to understand 😄 .

File: modules/InstallBasicPackageFiles.cmake [hunk 1]
```diff
@@ -142,216 +164,305 @@ include(CMakeParseArguments)
...
+         DEFINED BUILD_${_IBPF_VARS_PREFIX}_INCLUDEDIR OR
+         DEFINED ${_IBPF_VARS_PREFIX}_INSTALL_INCLUDEDIR OR
+         DEFINED INSTALL_${_IBPF_VARS_PREFIX}_INCLUDEDIR)
+        set(_get_include_dir "set(${_IBPF_VARS_PREFIX}_INCLUDEDIR \"\@PACKAGE_${_IBPF_VARS_PREFIX}_INCLUDEDIR\@\")")
+        set(_set_include_dir "set(${_Name}_INCLUDE_DIRS \${${_IBPF_VARS_PREFIX}_INCLUDEDIR})")
+      elseif(DEFINED ${_IBPF_VARS_PREFIX}_BUILD_INCLUDE_DIR OR
+             DEFINED BUILD_${_IBPF_VARS_PREFIX}_INCLUDE_DIR OR
+             DEFINED ${_IBPF_VARS_PREFIX}_INSTALL_INCLUDE_DIR OR
+             DEFINED INSTALL_${_IBPF_VARS_PREFIX}_INCLUDE_DIR)
+        string(APPEND _get_include_dir "set(${_IBPF_VARS_PREFIX}_INCLUDE_DIR \"\@PACKAGE_${_IBPF_VARS_PREFIX}_INCLUDE_DIR\@\")")
```
Comment (claudiofantacci) [inline modules/InstallBasicPackageFiles.cmake]: Shouldn't this and the following line a `set()` instead of a `string(APPEND ...)`?

Comment (drdanz) [reply modules/InstallBasicPackageFiles.cmake]: Thanks! Actually string(APPEND) does not exist until CMake 3.4, so you found a big compatibility issue 👍

Comment (drdanz) [reply modules/InstallBasicPackageFiles.cmake]: Repalced all `string(APPEND)`
Meanwhile I also fixed some vicious bug when exporting multiple targets.

**Comment by claudiofantacci** [comment]
I'll try to test this asap, I'm in the middle of several deliverables atm.

**Comment by traversaro** [comment]
I was not aware of the new `DEPENDENCIES` options, that's great!

**Pull request merged**